### PR TITLE
Week 5 Calendar Project

### PR DESCRIPTION
**Pull request by SydneyCaulfeild**
Create a program that returns a list of available meeting times for a group of attendees. The program takes into consideration existing meetings that these attendees may have.

### CODE REVIEW

**Review by Vector-S — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,117 @@
...
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    int requestDuration = (int) request.getDuration();
+
+    //if there are no events and the request duration is less than a day, return the whole day
```
Comment (lipczak-g) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:29]: nice! Good check if you expect most of the calendars to be sparse

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,117 @@
...
+    if (events.isEmpty() && requestDuration <= TimeRange.WHOLE_DAY.duration()){
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    //if the requested meeting is longer than a day, return an empty list since this is not possible
+    if (requestDuration > TimeRange.WHOLE_DAY.duration()) {
+        return Collections.emptyList();
+    }
+
+    //an arraylist to hold an ordered list without duplicates of all busy times in a day
```
Comment (lipczak-g) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:38]: nit: what is the order?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,117 @@
...
+    ArrayList<TimeRange> busyTimes = new ArrayList<TimeRange>();
+
+    //adding the beginning and end of the day as endpoints (your meeting can't be before the beginning of the day or after the end of the day)
+    busyTimes.add(TimeRange.fromStartDuration(0, 0));
+    busyTimes.add(TimeRange.fromStartDuration(1440,0));
```
Comment (lipczak-g) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:43]: nit: prefer 60 * 24 to make it clear to the reader

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,117 @@
...
+
+    for (Event event : events){
+        //create a timerange for the event we're evaluating
+        TimeRange newEvent = TimeRange.fromStartEnd(event.getWhen().start(), event.getWhen().duration()+event.getWhen().start(), false);
+        int newEventStart = newEvent.start();
+        int newEventEnd = newEvent.end();
+
+        //if none of the attendees in the request are attending this event, don't add it to BusyTimes because it's not relevant
+        Set<String> eventAttendees = event.getAttendees();
```
Comment (lipczak-g) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:52]: both variables are unnecessary, just add them to the disjoint call directly.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,117 @@
...
+        Collection<String> requestAttendees = request.getAttendees();
+        if (Collections.disjoint(eventAttendees, requestAttendees)){
+            continue;
```
Comment (lipczak-g) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:55]: I like that - it is better to quit fast rather than adding complex code to the if statement

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,117 @@
...
+        for (int i = 0; i < busyTimes.size()-1; i++) {
+            int currentBusyStart = busyTimes.get(i).start();
+            int currentBusyEnd = busyTimes.get(i).end();
+            int nextBusyStart = busyTimes.get(i+1).start();
+            int nextBusyEnd = busyTimes.get(i+1).end();
+            int adjustedNewStart;
+            int adjustedNewEnd;
+
+            //if the current busy time contains the new event, or they are equal, do not add the new event to the arraylist as it already exists
+            /*****/  //current busy block
```
Comment (lipczak-g) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:67]: lol, I spend a minute trying to understand what is going on here and why it works, please try to avoid confusion by not using comment markers for anything else, here is an example alternative

// current busy block: #######
// new event:                  ####

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,117 @@
...
+             /**/    //new event
+            if (busyTimes.get(i).contains(newEvent) || busyTimes.get(i).equals(newEvent)){
+                break; 
+            }
+            //if the new event starts during the current busy time and ends before the next busy time
+            /*******/        /**/       //busy blocks
+                 /******/               //new event
+            else if (TimeRange.contains(busyTimes.get(i), newEventStart) && newEventEnd <= nextBusyStart) {
```
Comment (lipczak-g) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:75]: This overall logic is complicated, hard to follow and (I think) missing some edge cases (e.g. what if the new event overlaps with multiple events). I think a better alternative would be to
1. find ALL events that overlap with the new event
2. replace them with new event that for which startTime = min(all affected events start time) and endTime = max(all affected events end time)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,117 @@
...
+                adjustedNewStart = newEventStart;
+                adjustedNewEnd = nextBusyStart;
+            }
+            //the new event does not start during the current busy block or before the next busy block, so move on to next busy block
+            else {
+                continue;
+            }
+
+            //add the new busy time with the adjusted start and duration
+            busyTimes.add(TimeRange.fromStartEnd(adjustedNewStart, adjustedNewEnd, false));
```
Comment (lipczak-g) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:106]: by adding and not replacing you will stack overlapping events and your logic of looking at i-th and (i+1)-th event will not work properly.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,117 @@
...
+            //sort the list in chronological order by event start time
+            Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
```
Comment (lipczak-g) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:108]: sorting is costly! I'd rather you add events at the end, iterate over all of them each time and then sort everything at the end. As I pointed out above there are edge cases in which checking the two neighbors is not sufficient anyway. Alternatively you can try to use a TreeSet which sorts on insert.
